File: exercises/posix-thead/exercise04.cpp
// =================================================================
//
// File: exercise04.cpp
// Author(s): Esteban Padilla Cerdio - A01703068
//            Hilda Olivia Beltrán Acosta - A01251916
// Description: This file implements the PI approximation using the
//				series proposed by Euler using using pthreads.
//              To compile: g++ exercise04.cpp -lpthread
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include "utils.h"
#include <cmath>
#include <iomanip>
#include <iostream>
#include <omp.h>
#include <pthread.h>

const int LIMIT = 1000000; // 1e6

using namespace std;

const int THREADS = 20;

// implement your code here

typedef struct
{
	long double start, end; // [start, end)
} Block;

void *calcPI(void *param)
{
	long double *acum;
	Block *block;

	block = (Block *)param;
	acum = new long double;
	(*acum) = 0;
	for (long double i = block->start; i < block->end; i++)
	{
		(*acum) += 1.0 / (i * i);
	}

	return ((void **)acum);
}

int main(int argc, char *argv[])
{
	int block_size, i, j;
	double ms;
	long double result, *acum;
	Block blocks[THREADS];
	pthread_t tids[THREADS];

	block_size = LIMIT / THREADS;

	for (i = 0; i < THREADS; i++)
	{
		blocks[i].start = i * block_size + 1;
		if (i != (THREADS - 1))
		{
			blocks[i].end = (i + 1) * block_size;
		}
		else
		{
			blocks[i].end = LIMIT;
		}
	}

	cout << "Starting..." << endl;
	ms = 0;

	for (j = 0; j < NUM; j++)
	{
		start_timer();

		result = 0;
		for (i = 0; i < THREADS; i++)
		{
			pthread_create(&tids[i], NULL, calcPI, (void *)&blocks[i]);
		}
		for (i = 0; i < THREADS; i++)
		{
			pthread_join(tids[i], (void **)&acum);
			result += (*acum);
			delete acum;
		}

		result = sqrt(6.0 * result);
		ms += stop_timer();
	}

	cout << "result = " << fixed << setprecision(10) << result << "\n";
	cout << "avg time = " << setprecision(5) << (ms / NUM) << " ms" << endl;

	return 0;
}

File: exercises/processes-python/exercise04.py
# =================================================================
##
# File: exercise04.py
# Author(s): Esteban Padilla Cerdio - A01703068
# 			 Hilda Olivia Beltrán Acosta - A01251916
# Description: This file implements the PI approximation using the
# series proposed by Euler.
# pi = sqrt ( 6 * sumatoria(i = 1-N) (1 / i^2) )
# using processes in Python.
##
# Copyright (c) 2022 by Tecnologico de Monterrey.
# All Rights Reserved. May be reproduced for any non-commercial
# purpose.
##
# =================================================================

import utils
import time
import multiprocessing as mp
import threading

SIZE = 1000000  # 1e6


def partialSum(start, end, queue):
    sum = 0
    for i in range(start, end):
        sum += 1 / (i ** 2)
    queue.put(sum)


if __name__ == "__main__":

    print("Starting...")

    CORES = 10

    startTime = endTime = ms = 0
    blockSize = SIZE // CORES

    for i in range(utils.N):
        startTime = time.time() * 1000

        queue = mp.SimpleQueue()

        processes = list()

        for i in range(CORES):
            start = i * blockSize + 1
            if i != (CORES - 1):
                end = (i + 1) * blockSize
            else:
                end = SIZE

            processes.append(mp.Process(
                target=partialSum, args=(start, end, queue)))
            processes[i].start()

        for i in range(CORES):
            processes[i].join()

        result = 0
        for i in range(CORES):
            result += queue.get()

        result = (6 * result) ** 0.5

        endTime = time.time() * 1000

        ms = ms + (endTime - startTime)

    print("result = ", result)
    print(CORES, "avg time = ", (ms / utils.N), " ms")
